native-codec/app/src/main/java/com/example/nativecodec/NativeCodec.java

```diff
@@ -316,4 +333,38 @@ void useAsSinkForNative() {
 
     }
 
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        if (NATIVE_CODEC_REQUEST != requestCode) {
+            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+            return;
+        }
+
+        if (grantResults.length != 1  ||
+                grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+            /*
+             * When user denied the permission, throw a Toast to prompt that READ_EXTERNAL_STORAGE
+             * is necessary;
+             * This application go back to the original state: it behaves as if the button
+             * was not clicked. The assumption is that user will re-click the "start" button
+             * (to retry), or shutdown the app in normal way.
+             */
+            Toast.makeText(getApplicationContext(),
+                    getString(R.string.NeedPermission),
```

Comment:
You can reference string resources directly in Toast.makeText, no need to call 'getString': http://developer.android.com/reference/android/widget/Toast.html#makeText(android.content.Context, int, int)